src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerDiagnosticListener.cs

```diff
@@ -67,49 +67,67 @@ public HttpHandlerDiagnosticListener(HttpClientInstrumentationOptions options)
 
         public override void OnStartActivity(Activity activity, object payload)
         {
+            // The overall flow of what HttpClient library does is as below:
+            // Activity.Start()
+            // DiagnosticSource.WriteEvent("Start", payload)
+            // DiagnosticSource.WriteEvent("Stop", payload)
+            // Activity.Stop()
+
+            // This method is in the WriteEvent("Start", payload) path.
+            // By this time, samplers have already run and
+            // activity.IsAllDataRequested populated accordingly.
+
+            if (Sdk.SuppressInstrumentation)
+            {
+                return;
+            }
+
             if (!this.startRequestFetcher.TryFetch(payload, out HttpRequestMessage request) || request == null)
             {
                 HttpInstrumentationEventSource.Log.NullPayload(nameof(HttpHandlerDiagnosticListener), nameof(this.OnStartActivity));
                 return;
             }
 
+            // TODO: Investigate why this check is needed.
```

Comment:
I think it was put for redirection cases where a redirect request might already have the context in the headers.

But I don't think the listener callback gets called for redirect cases. I tried this scenario: Call `http://google.com` from an HttpClient. This results in a 301 redirection to `http:www.//google.com`. The way HttpHandlerDiagnostics is currently configured, we only see one `OnStartActivity` and one `OnStopActivity` method call. In the `OnStartActivity` method they request payload has the original Uri: `http://google.com` whereas in the `OnStopActivity` method the payload has the redirected Uri: `http:www.//google.com`. This would mean that we only enter `OnStartActivity` once for the original HttpRequestOut call. So this logic would not be used for redirects.

Reply:
I know there were cases when I was working on HttpWebRequest instrumentation where it would re-use the request, but it has been a while so my memory is foggy. I think SSL upgrade/downgrade was one of them? Pretty sure I added a test or two for it. That is for the .NET Framework stuff but posting here because there are probably many similarities with the .NET Core side.